assets/js/googlesitekit/modules/datastore/settings.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * `core/modules` data store settings
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+const { createRegistrySelector } = Data;
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+export const actions = {
+	/**
+	 * Submits all changes for a module.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string} slug Slug for module store.
+	 * @return {Object} Module's submitChanges response object if it exists, otherwise object with `error` property if it doesn't.
+	 */
+	*submitChanges( slug ) {
+		invariant( slug, 'slug is required.' );
```

Comment:
I'm not sure if this is still an issue, due to a recent change in the way the tests run, but using `invariant` in a generator like this wouldn't cause the test to fail due to the asynchronous nature of how these run in the datastore.

The best way to check here would be to add a test to dispatch `submitChanges` without a slug and without `await` and expect it to throw the invariant message. E.g. https://github.com/google/site-kit-wp/blob/c042f538e0e6e97de9d9eac7453d2f01b435f52e/assets/js/googlesitekit/data/create-settings-store.test.js#L77-L80

If that doesn't work, then we simply need to change `submitChanges` here to not be a generator, but return one. E.g. https://github.com/google/site-kit-wp/blob/c042f538e0e6e97de9d9eac7453d2f01b435f52e/assets/js/googlesitekit/data/create-fetch-store.js#L177-L186
I thought we had more examples of this but I couldn't find one. It may be that they were refactored and this technique was no longer necessary.

Reply:
Approach 1 won't work.
With approach 2, it feels somewhat less correct to change the implementation so as to accommodate for the testing scenario, as it becomes less readable and thus less maintable, what if we instead:

`*submitChanges( slug ) {
		slug = slug ? slug : '';`

Which will correctly error on submitChanges `await registry.dispatch( STORE_NAME ).submitChanges()` with the following `{ error: "'modules/' does not have a submitChanges() action." };`

Reply:
> With approach 2, it feels somewhat less correct to change the implementation so as to accommodate for the testing scenario, as it becomes less readable and thus less maintable, what if we instead

It's not entirely for testing, and we don't actually need to split the two up as much as is done in the create fetch store example. This will work just as well:
```js
submitChanges( slug ) {
	invariant( slug, 'slug is required.' );

	return ( function* () {
		// yield ...
		// yield ...
	} )();
```

To your point however, in this case I think it would be fine to omit the invariant call and let it return the error as you suggested. I don't think it's necessary to have the fallback value though.

Reply:
Updated, and removed the fallback.

Reply:
Updated to include the invariant with generator in the fn.